proposals/interface-types/working-notes/scenarios/chooser.md

```diff
@@ -0,0 +1,680 @@
+# Choosing Strings
+
+## Preamble
+This note is part of a series of notes that gives complete end-to-end examples
+of using Interface Types to export and import functions.
+
+## Introduction
+
+Keeping track of memory that is allocated in order to allow string processing
+can be challenging. This example illustrates an extreme case that involves
+non-determinism.
+
+This scenario is based on the idea of using run-time information in order to
+control the flow of information. In particular, consider the C++ chooser
+function:
+
+```C++
+typedef std::shared_ptr<std::string> shared_string;
+
+shared_string nondeterministic_choice(shared_string one, shared_string two) {
```

Comment:
This is an odd C++ function. std::move mixed with shared_ptr is a weird one. It makes it unclear whether we're trying to re-use the existing shared_ptr without incrementing the refcounts, and if so it's unclear why we're using shared_ptr instead of directly passing a `std::string&&`.

This sample feels overloaded, like it's trying to illustrate three things at once. At least it needs to solve three things:

1. how to handle C++ shared_ptrs
2. how to reuse a passed-in reference (simpler version: identity function)
3. how to handle conditional ownership

It would be easier to follow and discuss if these were more explicitly separated. As it is now, they seem subtly interleaved, and it's hard to disentangle problems with shared_ptr vs problems with ownership.

Reply:
The reason for choosing the chooser is to highlight the necessity of the machinery. We start with a non-deterministic scenario and show how to implement it.
However, I defer to others wrt C++ style. What I was trying to do was build a scenario where the input may come back (or not) in the return result, and where copying would not be an acceptable answer.